internal/cri/server/container_image_mount.go

```diff
@@ -0,0 +1,213 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package server
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	containerd "github.com/containerd/containerd/v2/client"
+	"github.com/containerd/containerd/v2/core/leases"
+	"github.com/containerd/containerd/v2/core/mount"
+	"github.com/containerd/containerd/v2/defaults"
+	"github.com/containerd/errdefs"
+	"github.com/containerd/platforms"
+	"github.com/opencontainers/image-spec/identity"
+	imagespec "github.com/opencontainers/image-spec/specs-go/v1"
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+const (
+	labelGCSnapRef = "containerd.io/gc.ref.snapshot."
+)
+
+func (c *criService) mutateMounts(
+	ctx context.Context,
+	extraMounts []*runtime.Mount,
+	sandboxID string,
+	snapshotter string,
+	platform imagespec.Platform,
+
+) error {
+	for _, m := range extraMounts {
+		err := c.mutateImageMount(ctx, m, sandboxID, snapshotter, platform)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *criService) mutateImageMount(
+	ctx context.Context,
+	extraMount *runtime.Mount,
+	sandboxID string,
+	snapshotter string,
+	platform imagespec.Platform,
+) (retErr error) {
+	imageSpec := extraMount.GetImage()
+	if imageSpec == nil {
+		return nil
+	}
+	if extraMount.GetHostPath() != "" {
+		return fmt.Errorf("hostpath must be empty while mount image: %+v", extraMount)
+	}
+	if !extraMount.GetReadonly() {
+		return fmt.Errorf("readonly must be true while mount image: %+v", extraMount)
+	}
+
+	ref := imageSpec.GetImage()
+	if ref == "" {
+		return fmt.Errorf("image not specified in: %+v", imageSpec)
+	}
+	image, err := c.LocalResolve(ref)
+	if err != nil {
+		return fmt.Errorf("failed to resolve image %q: %w", ref, err)
+	}
+	containerdImage, err := c.toContainerdImage(ctx, image)
+	if err != nil {
+		return fmt.Errorf("failed to get image from containerd %q: %w", image.ID, err)
+	}
+
+	// This is a digest of the manifest
+	imageID := containerdImage.Target().Digest.Hex()
+
+	target := c.getImageVolumeHostPath(sandboxID, imageID)
+
+	// Already mounted in another container on the same pod
+	if stat, err := os.Stat(target); err == nil && stat.IsDir() {
+		extraMount.HostPath = target
+		return nil
+	}
+
+	ctx, done, err := c.client.WithLease(ctx,
```

Comment:
@dmcgowan Should the snapshot be owned by the sandbox rather than held by a lease?  Seems like that would handle GC better.

Reply:
Yes, the reference label should be put on the sandbox or the container

Reply:
sandbox is probably the right answer as we don't want to gc the image until the sandbox is no more.. in case the container is restarted.. or if there are N containers using the same image volume... Later we can explore sharing the volume across a set of sandboxes.  But at this point in time there were test buckets that tried to verify the image volume is removed after the sandbox is removed...

Reply:
> Later we can explore sharing the volume across a set of sandboxes

That's already handled by the nature of snapshots, right? We'll have one set of snapshots per image (roughly 1:1 layer:snapshot). If there are multiple references to the same image, each held by a different sandbox, the image and its snapshots won't be GC'd until all the references (sandboxes) are removed.

Reply:
Nod the oci volume image pulled by kubelet may be gc'd per kubelet procedures for reducing images stored on the node and thus layers no longer shared across images could also be collected, but the host volume path and contents created from the image and image config has the life cycle of the pod once created and subsequently bind mounted for a target path in at least one of the pod's containers.. we could change the prefix  ^ for the host path of the volume to be imageref only based vs podid+imageref.. but it is a slightly different pattern than we use today for container images.